File: Emulator/src/main/core/game/world/map/zone/MapZone.java
package core.game.world.map.zone;

import content.global.skill.summoning.familiar.Familiar;
import core.game.interaction.Option;
import core.game.node.Node;
import core.game.node.entity.Entity;
import core.game.node.entity.combat.CombatStyle;
import core.game.node.entity.npc.NPC;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.request.RequestType;
import core.game.node.item.Item;
import core.game.world.map.Location;
import core.game.world.map.Region;
import core.game.world.map.RegionManager;

import java.util.Iterator;
import java.util.Objects;

/**
 * The type Map zone.
 */
public abstract class MapZone implements Zone {

    private int uid;

    private String name;

    private boolean overlappable;

    /**
     * The Fire random events.
     */
    protected boolean fireRandomEvents;

    private int restriction;

    private int zoneType;

    /**
     * Instantiates a new Map zone.
     *
     * @param name         the name
     * @param overlappable the overlappable
     * @param restrictions the restrictions
     */
    public MapZone(String name, boolean overlappable, ZoneRestriction... restrictions) {
        this.name = name;
        this.overlappable = overlappable;
        for (ZoneRestriction restriction : restrictions) {
            addRestriction(restriction.getFlag());
        }
    }

    @Override
    public boolean enter(Entity e) {
        if (e instanceof Player) {
            Player p = (Player) e;
        } else if (e instanceof NPC) {
            NPC npc = (NPC) e;
            if (e instanceof Familiar && isRestricted(ZoneRestriction.FOLLOWERS.getFlag())) {
                npc.setInvisible(true);
            }
        }
        return true;
    }

    @Override
    public boolean leave(Entity e, boolean logout) {
        return true;
    }

    /**
     * Can logout boolean.
     *
     * @param p the p
     * @return the boolean
     */
    public boolean canLogout(Player p) {
        return true;
    }

    /**
     * Death boolean.
     *
     * @param e      the e
     * @param killer the killer
     * @return the boolean
     */
    public boolean death(Entity e, Entity killer) {
        return false;
    }

    /**
     * Interact boolean.
     *
     * @param e      the e
     * @param target the target
     * @param option the option
     * @return the boolean
     */
    public boolean interact(Entity e, Node target, Option option) {
        return false;
    }

    /**
     * Handle use with boolean.
     *
     * @param player the player
     * @param used   the used
     * @param with   the with
     * @return the boolean
     */
    public boolean handleUseWith(Player player, Item used, Node with) {
        return false;
    }

    /**
     * Action button boolean.
     *
     * @param player      the player
     * @param interfaceId the interface id
     * @param buttonId    the button id
     * @param slot        the slot
     * @param itemId      the item id
     * @param opcode      the opcode
     * @return the boolean
     */
    public boolean actionButton(Player player, int interfaceId, int buttonId, int slot, int itemId, int opcode) {
        return false;
    }

    /**
     * Continue attack boolean.
     *
     * @param e       the e
     * @param target  the target
     * @param style   the style
     * @param message the message
     * @return the boolean
     */
    public boolean continueAttack(Entity e, Node target, CombatStyle style, boolean message) {
        return true;
    }

    /**
     * Ignore multi boundaries boolean.
     *
     * @param attacker the attacker
     * @param victim   the victim
     * @return the boolean
     */
    public boolean ignoreMultiBoundaries(Entity attacker, Entity victim) {
        return false;
    }

    /**
     * Check multi boolean.
     *
     * @param e       the e
     * @param t       the t
     * @param message the message
     * @return the boolean
     */
    public static boolean checkMulti(Entity e, Entity t, boolean message) {
        long time = System.currentTimeMillis();
        boolean multi = t.getProperties().isMultiZone() && e.getProperties().isMultiZone();
        if (multi || e.isIgnoreMultiBoundaries(t) || e.getZoneMonitor().isIgnoreMultiBoundaries(t)) {
            return true;
        }
        Entity target = t.getAttribute("combat-attacker", e);
        if (t.getAttribute("combat-time", -1L) > time && target != e && target.isActive()) {
            if (message && e instanceof Player) {
                ((Player) e).getPacketDispatch().sendMessage("Someone else is already fighting this" + (t instanceof Player ? " player." : "."));
            }
            return false;
        }
        if (e.getAttribute("combat-time", -1L) > time && (target = e.getAttribute("combat-attacker", t)) != t && target.isActive()) {
            if (t.getId() == 1614 || t.getId() == 1613) {
                return true;
            }
            if (message && e instanceof Player) {
                ((Player) e).getPacketDispatch().sendMessage("You're already under attack!");
            }
            return false;
        }
        return true;
    }

    /**
     * Teleport boolean.
     *
     * @param e    the e
     * @param type the type
     * @param node the node
     * @return the boolean
     */
    public boolean teleport(Entity e, int type, Node node) {
        return true;
    }

    /**
     * Start death boolean.
     *
     * @param e      the e
     * @param killer the killer
     * @return the boolean
     */
    public boolean startDeath(Entity e, Entity killer) {
        return true;
    }

    /**
     * Can request boolean.
     *
     * @param type   the type
     * @param player the player
     * @param target the target
     * @return the boolean
     */
    public boolean canRequest(RequestType type, Player player, Player target) {
        return true;
    }

    /**
     * Move boolean.
     *
     * @param e    the e
     * @param from the from
     * @param to   the to
     * @return the boolean
     */
    public boolean move(Entity e, Location from, Location to) {
        return true;
    }

    /**
     * Parse command boolean.
     *
     * @param player    the player
     * @param name      the name
     * @param arguments the arguments
     * @return the boolean
     */
    public boolean parseCommand(Player player, String name, String[] arguments) {
        return false;
    }

    /**
     * Location update.
     *
     * @param e    the e
     * @param last the last
     */
    public void locationUpdate(Entity e, Location last) {

    }

    /**
     * Configure.
     */
    public void configure() {
    }

    ;

    /**
     * Clean items.
     *
     * @param player the player
     * @param items  the items
     */
    public void cleanItems(Player player, Item[] items) {
        if (player == null) {
            return;
        }
        for (Item item : items) {
            if (item == null) {
                continue;
            }
            if (player.getInventory().containsItem(item)) {
                player.getInventory().remove(new Item(item.getId(), player.getInventory().getAmount(item)));
            }
            if (player.getEquipment().containsItem(item)) {
                player.getEquipment().remove(new Item(item.getId(), player.getEquipment().getAmount(item)));
            }
            if (player.getBank().containsItem(item)) {
                player.getBank().remove(new Item(item.getId(), player.getBank().getAmount(item)));
            }
        }
    }

    /**
     * Message.
     *
     * @param e       the e
     * @param message the message
     */
    protected static void message(Entity e, String message) {
        if (!(e instanceof Player)) {
            return;
        }
        ((Player) e).getPacketDispatch().sendMessage(message);
    }

    /**
     * Register.
     *
     * @param borders the borders
     */
    public void register(ZoneBorders borders) {
        for (Integer id : borders.getRegionIds()) {
            Region r = RegionManager.forId(id);
            if (r != null) {
                r.add(new RegionZone(this, borders));
            }
        }
    }

    /**
     * Unregister.
     *
     * @param borders the borders
     */
    public void unregister(ZoneBorders borders) {
        for (Integer id : borders.getRegionIds()) {
            Region r = RegionManager.forId(id);
            if (r != null) {
                r.remove(new RegionZone(this, borders));
            }
        }
    }

    /**
     * Register region.
     *
     * @param regionId the region id
     */
    public void registerRegion(int regionId) {
        register(ZoneBorders.forRegion(regionId));
    }

    /**
     * Register region.
     *
     * @param regionId the region id
     * @param borders  the borders
     */
    public void registerRegion(int regionId, ZoneBorders borders) {
        Region r = RegionManager.forId(regionId);
        if (r != null) {
            r.add(new RegionZone(this, borders));
        }
    }

    /**
     * Unregister region.
     *
     * @param regionId the region id
     */
    public void unregisterRegion(int regionId) {
        Region r = RegionManager.forId(regionId);
        if (r != null) {
            for (Iterator<RegionZone> it = r.getRegionZones().iterator(); it.hasNext(); ) {
                if (it.next().getZone() == this) {
                    it.remove();
                }
            }
        }
    }

    /**
     * Disable random events.
     */
    public void disableRandomEvents() {
        this.fireRandomEvents = false;
    }

    /**
     * Gets name.
     *
     * @return the name
     */
    public String getName() {
        return name;
    }

    /**
     * Sets name.
     *
     * @param name the name
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * Is overlappable boolean.
     *
     * @return the boolean
     */
    public boolean isOverlappable() {
        return overlappable;
    }

    /**
     * Sets overlappable.
     *
     * @param overlappable the overlappable
     */
    public void setOverlappable(boolean overlappable) {
        this.overlappable = overlappable;
    }

    /**
     * Gets uid.
     *
     * @return the uid
     */
    public int getUid() {
        return getName().hashCode();
    }

    /**
     * Sets uid.
     *
     * @param uid the uid
     */
    public void setUid(int uid) {
        this.uid = uid;
    }

    /**
     * Is fire randoms boolean.
     *
     * @return the boolean
     */
    public boolean isFireRandoms() {
        return fireRandomEvents;
    }

    /**
     * Is dynamic zone boolean.
     *
     * @return the boolean
     */
    public boolean isDynamicZone() {
        return false;
    }

    /**
     * Add restriction.
     *
     * @param restriction the restriction
     */
    public void addRestriction(ZoneRestriction restriction) {
        addRestriction(restriction.getFlag());
    }

    /**
     * Add restriction.
     *
     * @param flag the flag
     */
    public void addRestriction(int flag) {
        restriction |= flag;
    }

    /**
     * Is restricted boolean.
     *
     * @param flag the flag
     * @return the boolean
     */
    public boolean isRestricted(int flag) {
        return (restriction & flag) != 0;
    }

    /**
     * Gets restriction.
     *
     * @return the restriction
     */
    public int getRestriction() {
        return restriction;
    }

    /**
     * Gets zone type.
     *
     * @return the zone type
     */
    public int getZoneType() {
        return zoneType;
    }

    /**
     * Sets zone type.
     *
     * @param type the type
     */
    public void setZoneType(int type) {
        this.zoneType = type;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MapZone mapZone = (MapZone) o;
        return getUid() == mapZone.getUid();
    }

    @Override
    public int hashCode() {
        return Objects.hash(uid, name, overlappable, fireRandomEvents, restriction, zoneType);
    }
}

File: Server/src/main/core/game/world/map/zone/MapZone.kt
package core.game.world.map.zone

import content.global.skill.summoning.familiar.Familiar
import core.game.interaction.Option
import core.game.node.Node
import core.game.node.entity.Entity
import core.game.node.entity.combat.CombatStyle
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.entity.player.link.request.RequestType
import core.game.node.item.Item
import core.game.world.map.Location
import core.game.world.map.Region
import core.game.world.map.RegionManager
import java.util.Objects

/**
 * Represents a map zone.
 * @author Emperor
 */
abstract class MapZone(
    /**
     * The name of this map zone.
     */
    var name: String,

    /**
     * If the map zone can be overlapped by another zone.
     */
    var overlappable: Boolean,

    /**
     * Optional restrictions to apply to this zone.
     */
    vararg restrictions: ZoneRestriction
) : Zone {

    /**
     * The map zone UID.
     */
    var uid: Int = 0

    /**
     * If random events should be fired in this zone.
     */
    protected var fireRandomEvents: Boolean = true

    /**
     * Restriction flags.
     */
    private var restriction: Int = 0

    /**
     * Zone type (used for items kept on death).
     */
    private var zoneType: Int = 0

    init {
        restrictions.forEach { addRestriction(it.flag) }
    }

    /**
     * Called when an entity enters this zone.
     * @param e The entity entering.
     * @return True if entry was successful.
     */
    override fun enter(e: Entity): Boolean {
        when (e) {
            is Player -> {}
            is NPC -> if (e is Familiar && isRestricted(ZoneRestriction.FOLLOWERS.flag)) {
                e.isInvisible = true
            }
        }
        return true
    }

    /**
     * Called when an entity leaves this zone.
     * @param e The entity leaving.
     * @param logout If the entity is logging out.
     * @return True if leaving is allowed.
     */
    override fun leave(e: Entity, logout: Boolean): Boolean = true

    /**
     * Checks if the player can log out in this zone.
     */
    open fun canLogout(p: Player): Boolean = true

    /**
     * Called when an entity dies in this zone.
     */
    open fun death(e: Entity, killer: Entity): Boolean = false

    /**
     * Handles interaction with a target node.
     */
    open fun interact(e: Entity, target: Node, option: Option): Boolean = false

    /**
     * Handles using an item with a node.
     */
    open fun handleUseWith(player: Player, used: Item, with: Node): Boolean = false

    /**
     * Handles an action button click.
     */
    open fun actionButton(player: Player, interfaceId: Int, buttonId: Int, slot: Int, itemId: Int, opcode: Int): Boolean = false

    /**
     * Checks if the entity can continue attacking a target.
     */
    open fun continueAttack(e: Entity, target: Node, style: CombatStyle, message: Boolean): Boolean = true

    /**
     * Checks if multi-zone boundaries should be ignored for combat.
     */
    open fun ignoreMultiBoundaries(attacker: Entity, victim: Entity): Boolean = false

    companion object {
        /**
         * Checks if an entity can attack another entity respecting multi-zone rules.
         * @param e Attacker entity.
         * @param t Target entity.
         * @param message Whether to send messages.
         * @return True if attack can continue.
         */
        fun checkMulti(e: Entity, t: Entity, message: Boolean): Boolean {
            val time = System.currentTimeMillis()
            val multi = t.properties.isMultiZone && e.properties.isMultiZone
            if (multi || e.isIgnoreMultiBoundaries(t) || e.zoneMonitor.isIgnoreMultiBoundaries(t)) return true

            var target = t.getAttribute("combat-attacker", e)
            if (t.getAttribute("combat-time", -1L) > time && target != e && target.isActive) {
                if (message && e is Player) e.packetDispatch.sendMessage(
                    "Someone else is already fighting this${if (t is Player) " player." else "."}"
                )
                return false
            }

            val target2 = e.getAttribute("combat-attacker", t)
            if (e.getAttribute("combat-time", -1L) > time && target2 != t && target2.isActive) {
                if (t.id == 1614 || t.id == 1613) return true
                if (message && e is Player) e.packetDispatch.sendMessage("You're already under attack!")
                return false
            }
            return true
        }
    }

    /**
     * Checks if an entity can teleport in this zone.
     */
    open fun teleport(e: Entity, type: Int, node: Node?): Boolean = true

    /**
     * Called to start death sequence.
     */
    open fun startDeath(e: Entity, killer: Entity): Boolean = true

    /**
     * Checks if a request can be made in this zone.
     */
    open fun canRequest(type: RequestType, player: Player, target: Player): Boolean = true

    /**
     * Checks if an entity can move from one location to another.
     */
    open fun move(e: Entity, from: Location, to: Location): Boolean = true

    /**
     * Parses a command in this zone.
     */
    open fun parseCommand(player: Player, name: String, arguments: Array<String>): Boolean = false

    /**
     * Called when an entity changes location.
     */
    open fun locationUpdate(e: Entity, last: Location) {}

    /**
     * Configures this zone.
     */
    open fun configure() {}

    /**
     * Removes specified items from player inventory, equipment, and bank.
     * @param player The player.
     * @param items The items to remove.
     */
    fun cleanItems(player: Player?, items: Array<Item?>) {
        player ?: return
        items.forEach { item ->
            item ?: return@forEach
            if (player.inventory.containsItem(item)) player.inventory.remove(
                Item(
                    item.id, player.inventory.getAmount(item)
                )
            )
            if (player.equipment.containsItem(item)) player.equipment.remove(
                Item(
                    item.id, player.equipment.getAmount(item)
                )
            )
            if (player.bank.containsItem(item)) player.bank.remove(Item(item.id, player.bank.getAmount(item)))
        }
    }

    /**
     * Sends a message to a player entity.
     */
    protected fun message(e: Entity, message: String) {
        if (e !is Player) return
        e.packetDispatch.sendMessage(message)
    }

    /**
     * Registers this zone in the regions for the given borders.
     */
    fun register(borders: ZoneBorders) {
        borders.regionIds.forEach { id ->
            RegionManager.forId(id)?.add(RegionZone(this, borders))
        }
    }

    /**
     * Unregisters this zone from the regions for the given borders.
     */
    fun unregister(borders: ZoneBorders) {
        borders.regionIds.forEach { id ->
            RegionManager.forId(id)?.remove(RegionZone(this, borders))
        }
    }

    /**
     * Registers this zone in a region by id.
     */
    fun registerRegion(regionId: Int) = register(ZoneBorders.forRegion(regionId))

    /**
     * Registers this zone in a region with specific borders.
     */
    fun registerRegion(regionId: Int, borders: ZoneBorders) {
        RegionManager.forId(regionId)?.add(RegionZone(this, borders))
    }

    /**
     * Unregisters this zone from a region by id.
     */
    fun unregisterRegion(regionId: Int) {
        RegionManager.forId(regionId)?.regionZones?.removeIf { it.zone == this }
    }

    /**
     * Disables firing of random events in this zone.
     */
    fun disableRandomEvents() {
        fireRandomEvents = false
    }

    /**
     * Checks if random events are active.
     */
    fun isFireRandoms(): Boolean = fireRandomEvents

    /**
     * Checks if this is a dynamic zone (non-location based zones typically return true).
     */
    fun isDynamicZone(): Boolean = false

    /**
     * Adds a restriction by enum.
     */
    fun addRestriction(restriction: ZoneRestriction) {
        addRestriction(restriction.flag)
    }

    /**
     * Adds a restriction by flag.
     */
    fun addRestriction(flag: Int) {
        restriction = restriction or flag
    }

    /**
     * Checks if a restriction flag is active.
     */
    fun isRestricted(flag: Int): Boolean = restriction and flag != 0

    /**
     * Gets restriction flags.
     */
    fun getRestriction(): Int = restriction

    /**
     * Gets the zone type.
     */
    fun getZoneType(): Int = zoneType

    /**
     * Sets the zone type.
     */
    fun setZoneType(type: Int) {
        zoneType = type
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false
        other as MapZone
        return uid == other.uid
    }

    override fun hashCode(): Int = Objects.hash(uid, name, overlappable, fireRandomEvents, restriction, zoneType)

    /**
     * Computes the UID based on the zone name
     */
    fun getUid(): Int = name.hashCode()
}
